src/pool/mod.rs

```diff
@@ -8,6 +8,58 @@ use crate::{
 
 use std::{fmt, marker::PhantomData};
 
+/// A lock-free concurrent object pool.
+///
+/// Slabs provide pre-allocated storage for many instances of a single type. But when working with
+/// heap allocated objects, the advantages of a slab are lost as the memory allocated for the
+/// object is freed when the object is removed from the slab. With a pool, we can instead re-use
+/// this memory for objects being added to the pool in the future, therefore reducing memory
+/// fragmentation and avoiding additional allocations.
+///
+/// This type implements a lock-free concurrent pool, indexed by `usize`s. The items stored in this
+/// type need to implement [`Clear`] and `Default`.
+///
+/// The `Pool` type shares similar semantics to [`Slab`] when it comes to sharing across threads
+/// and storing mutable shared data. The biggest difference is there are no [`Slab::insert`] and
+/// [`Slab::take`] analouges for the `Pool` type. Instead new items are added to the pool by using
+/// the [`Pool::create`] method, and marked for clearing by the [`Pool::clear`] method.
+///
+/// # Examples
+///
+/// Add an entry to the pool, returning an index:
+/// ```
+/// # use sharded_slab::Pool;
+/// let pool: Pool<String> = Pool::new();
+///
+/// let key = pool.create(|item| item.push_str("hello world")).unwrap();
+/// assert_eq!(pool.get(key).unwrap(), String::from("hello world"));
+/// ```
+///
+/// Pool entries can be cleared either by manually calling [`Pool::clear`] or when the guard is
+/// dropped. The entry will be cleared when all the references to it are dropped.
```

Comment:
Dropping a guard doesn't automatically clear the entry. Only `Pool::clear` does. Dropping the guard just decrements the reference count, indicating that the entry _can_ be cleared.

Let's change this to:
```suggestion
/// Pool entries can be cleared by calling [`Pool::clear`]. This marks the entry to be
/// cleared when all guards referencing to it are dropped.
```